<h1 id="january-30-2011">January 30, 2011</h1>
<ul>
<li>Updated data for several TCGA cancer studies</li>
<li>Some small bug-fixes</li>
</ul>
<h1 id="december-22-2011">December 22, 2011</h1>
<ul>
<li><strong>Fourteen new <a href="http://cancergenome.nih.gov/">TCGA</a> cancer studies</strong>:  This includes complete data for TCGA Colorectal Carcinoma
and provisional data for thirteen other cancer types in the TCGA production pipeline.  Please note that data from these
thirteen new cancer types are provisional, not final and do not yet include mutation data.
As per NCI guidelines, preliminary mutation data cannot be redistributed until they have been validated.<br />
</li>
</ul>
<p><center><img alt="TCGA" src="http://cancergenome.nih.gov/PublishedContent/Images/SharedItems/Images/tcga_logo.png" /></center></p>
<ul>
<li><strong>Four new data types</strong>:<ul>
<li>Reverse-phase protein array (RPPA) data.</li>
<li>microRNA expression and copy-number (including support for multiple loci)</li>
<li>RNA-Seq based expression data.</li>
<li>log2 copy-number data.</li>
</ul>
</li>
<li>Updated TCGA GBM copy-number, expression, and methylation data.</li>
<li>New <strong>gene symbol validation</strong> service.  You can now use gene aliases and/or Entrez Gene IDs within your gene sets.</li>
<li><strong>Links to IGV</strong> for visualization of DNA copy-number changes.</li>
<li>Background information from the <a href="http://www.sanger.ac.uk/genetics/CGP/Census/">Sanger Cancer Gene Census</a>.</li>
<li>Two <strong>new <a href="tutorial.jsp">Tutorials</a></strong> to get you quickly started in using the portal.
<BR/>
<BR/></li>
</ul>
<h1 id="november-14-2011">November 14, 2011</h1>
<ul>
<li>New and <strong>improved mutation details</strong>, with sorting and filtering capabilities.</li>
<li>In collaboration with Bilkent University, we have added a <strong>new Network tab</strong> to our results pages.  The network tab enables users to visualize, analyze and filter cancer genomic data in the context of pathways and interaction networks derived from <a href="http://www.pathwaycommons.org">Pathway Commons</a>.</li>
</ul>
<p><center><img alt="GBM Network" src="images/previews/ova_network.png" /></center>
<BR/>
<BR/></p>
<h1 id="september-3-2011">September 3, 2011</h1>
<ul>
<li>You can now query across different cancer studies (feature available directly from the home page).</li>
<li>Our <a href="cgds_r.jsp">MATLAB CGDS Cancer Genomics Toolbox</a> is now available.  The toolbox enables you to download data from the cBio Portal, and import it directly into MATLAB.</li>
<li>The code for the cBio Portal has now been fully open sourced, and made available at <a href="http://code.google.com/p/cbio-cancer-genomics-portal/">Google Code</a>.  If you would like to join our open source efforts and make the portal even better, drop us an email.</li>
</ul>
<p><center><img alt="Cross Cancer Query" src="images/previews/cross_cancer.png" /></center>
<BR/>
<BR/></p>
<h1 id="march-2-2011">March 2, 2011</h1>
<p>New plotting features and other improvements:</p>
<ul>
<li>Correlation plots that show the relationship between different data types for individual genes.</li>
<li>Survival analysis - assess survival differences between altered and non-altered patient sets.</li>
<li>Updated <a href="cgds_r.jsp">R Package</a> with support for correlation plots and general improvements for retrieving and accessing data in R data frames.</li>
<li>The <a href="web_api.jsp">Web Interface</a> now supports basic clinical data, e.g. survival data.</li>
<li><a href="networks.jsp">Networks</a> for pathway analysis are now available for download.</li>
</ul>
<p><center><img alt="Survival Analysis" src="images/previews/ova_survival_60_percent.png" /></center>
<BR/>
<BR/></p>
<h1 id="december-15-2010">December 15, 2010</h1>
<p>Several new features, including:</p>
<ul>
<li>Redesigned and streamlined user interface, based on user feedback and usability testing.</li>
<li>Advanced support for gene-specific alterations.  For example, users
can now view mutations within TP53, and ignore copy number alterations, or
only view amplifications of EGFR, and ignore deletions.</li>
<li>Improved performance.</li>
<li><a href="faq.jsp">Frequently Asked Questions</a> document released.</li>
<li>Updated <a href="video.jsp">Video Tutorial</a>.
<BR/>
<BR/></li>
</ul>
<h1 id="november-4-2010">November 4, 2010</h1>
<ul>
<li>Enhanced <a href="faq.jsp#what-are-oncoprints">Oncoprints</a>, enabling users to quickly visualize genomic alterations across many cases.  Oncoprints now also work in all major browsers, including Firefox, Chrome, Safari, and Internet Explorer.</li>
<li>Official release of our <a href="web_api.jsp">Web Interface</a>, enabling programmatic access to all data.</li>
<li>Official release of our <a href="cgds_r.jsp">R Package</a>, enabling programmatic access to all data from the R platform for statistical computing.</li>
</ul>
<p><center><img alt="OncoPrints" src="images/previews/gbm_oncoprint.png" /></center></p>